Zend/Optimizer/zend_inference.c

```diff
@@ -3369,8 +3369,17 @@ static zend_always_inline zend_result _zend_update_type_info(
 			ce = zend_optimizer_get_class_entry_from_op1(script, op_array, opline);
 			if (ce) {
 				UPDATE_SSA_OBJ_TYPE(ce, 0, ssa_op->result_def);
+				/* New objects without constructors cannot escape. */
+				if (!ce->constructor) {
```

Comment:
Following your discussion with @bwoebi, I think this may be changed into
```
if (!ce->constructor && !ce->create_object && ce->handlers->get_constructor == zend_std_get_constructor) {
```